src/Prototype/Database.hs

```diff
@@ -91,23 +96,31 @@ getProfileAndList h namespace listname = do
 
 
 --------------------------------------------------------------------------------
-newTodoLists = newTVar Examples.todoLists
+newTodoLists :: STM (TVar (Map String [TVar TodoList]))
+newTodoLists = do
+  let f (k, v) = mapM newTVar v >>= return . (k,)
+  todoLists <- mapM f Examples.todoLists
+  newTVar (Map.fromList todoLists)
 
+getAllTodoLists :: Handle -> STM [TodoList]
 getAllTodoLists h = do
   lists <- readTVar (hTodoLists h)
-  return (map snd lists)
+  let lists' = Map.toList lists
+  mapM readTVar $ concat $ map snd lists'
 
+getTodoLists :: Handle -> String -> STM [TodoList]
 getTodoLists h namespace = do
   lists <- readTVar (hTodoLists h)
-  return ((map snd . filter f) lists)
-  where f = (namespace ==) . fst
+  case Map.lookup namespace lists of
+    Nothing -> return []
+    Just lists' -> mapM readTVar lists'
 
+getTodoList :: Handle -> String -> String -> STM (Maybe TodoList)
 getTodoList h namespace listname = do
-  lists <- readTVar (hTodoLists h)
-  case filter f lists of
-    [(_, list)] -> return (Just list)
+  lists <- getTodoLists h namespace
+  case filter ((listname ==) . tlName) lists of
```

Comment:
Maybe we should use: 
``` haskell 
pure $ case filter ... 
```

(Also pure instead of return, since we don't need the power of return per se) 

That would avoid the repeated `return`s in each match.